### PR TITLE
Enable nullable: System.Management.Automation.IContainsErrorRecord

### PR DESCRIPTION
**Pull request by powercode**
Tracking issue: #12631.

### CODE REVIEW

File: src/System.Management.Automation/engine/ErrorPackage.cs [hunk 1]
```diff
@@ -1790,6 +1792,8 @@ public interface IContainsErrorRecord
         ErrorRecord ErrorRecord { get; }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/ErrorPackage.cs:1792]:
```suggestion
        ErrorRecord? ErrorRecord { get; }
```

I found a lot of checks `ErrorRecord != null`. Some classes obviously return null. Some classes always return a value so we will annotate them appropriately.

**Review by iSazonov — CHANGES_REQUESTED** [review]

**Comment by powercode** [comment]
The checks I see are of the type
```csharp
 var icer = exception as IContainsErrorRecord;
if (icer == null)...
```

But that is just checking for the interface.

```csharp
        /// <summary>
        /// Encode exception.
        /// </summary>
        private static PSObject EncodeException(Exception exception)
        {
            // We are encoding exceptions as ErrorRecord objects because exceptions written
            // to the wire are lost during serialization. By sending across ErrorRecord objects
            // we are able to preserve the exception as well as the stack trace.

            ErrorRecord errorRecord = null;
            IContainsErrorRecord containsErrorRecord = exception as IContainsErrorRecord;
            if (containsErrorRecord == null)
            {
                // If this is a .NET exception then wrap in an ErrorRecord.
                errorRecord = new ErrorRecord(exception, "RemoteHostExecutionException", ErrorCategory.NotSpecified, null);
            }
            else
            {
                // Exception inside the error record is ParentContainsErrorRecordException which
                // doesn't have stack trace. Replace it with top level exception.
                errorRecord = containsErrorRecord.ErrorRecord;
                errorRecord = new ErrorRecord(errorRecord, exception);
            }

            PSObject errorRecordPSObject = RemotingEncoder.CreateEmptyPSObject();
            errorRecord.ToPSObjectForRemoting(errorRecordPSObject);
            return errorRecordPSObject;
        }
```

And 
```csharp
        public ErrorRecord(ErrorRecord errorRecord,
                             Exception replaceParentContainsErrorRecordException)
        {
            if (errorRecord == null)
            {
                throw new PSArgumentNullException(nameof(errorRecord));
            }
}
```

Clearly, the expectation is that it is non-null.

I don't think the checks you see for null are what you think they are.

File: src/System.Management.Automation/engine/ErrorPackage.cs (hunk 1 above)
Comment (powercode) [reply src/System.Management.Automation/engine/ErrorPackage.cs:1791]: No this is correct. 
If you implement IContainsErrorRecord, you can deliver one. 
Lots of code depends on that.

**Comment by powercode** [comment]
I think those checks are just extra precautions since we haven't had a language mechanism to help us with this.
So devs add safeguards.

**Review by vexx32 — APPROVED** [review]

File: src/System.Management.Automation/engine/ErrorPackage.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/ErrorPackage.cs:1791]: First link I opened was "_errorRecord = (ErrorRecord)info.GetValue("errorRecord", typeof(ErrorRecord));" but now I reviewed other implementations and agree with the design.

#closed.

File: src/System.Management.Automation/engine/ErrorPackage.cs [hunk 2]
```diff
@@ -1717,6 +1717,8 @@ public ErrorRecord(Exception exception, string errorId, ErrorCategory errorCateg
         }
     }
 
+#nullable enable
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/ErrorPackage.cs:1720]: @powercode Please move the directive to line 1764 before "public interface IContainsErrorRecord".
It is more reliable pattern for auto removing (not all types have XML comments)/

**Review by iSazonov — APPROVED** [review]

**Pull request merged**